File: dbscan.cpp
#include "dbscan.h"
#include<random>
#include<boost/algorithm/string.hpp>
#include<fstream>
#include<boost/lexical_cast.hpp>
#include<sstream>

using namespace boost;

Info::Info():
    id_(-1),
    group_(-1),
    visited_(false),
    x_(0.0),
    y_(0.0)
{

}

Info::Info(int id, string value, double x, double y){
    id_ = id;
    group_ = -1;
    value_ = value;
    visited_ = false;
    x_ = x;
    y_ = y;
}

void Info::setData(int id, int group, string value, bool visited, double x, double y){
    id_ = id;
    group_ = group;
    value_ = value;
    visited_ = visited;
    x_ = x;
    y_ = y;
    }

void Info::print(){
    std::cout << "id_: " << id_ << ", group_: " << group_ << ", value_: " << value_ << ", visited: " << visited_ << ", x_: " << x_ << ", y_: " << y_ << std::endl;
}


DBscan::DBscan( vector<Info> info){
    info_ = info;
}

vector<Info> DBscan::getInfo(){
    return info_;
}

void DBscan::grouping(map<int, vector<Info> > &vec){
    map<int, vector<Info> >::iterator it;
    for(int i = 0; i < info_.size(); i++){
        it = vec.find(info_[i].group_);
        if(it != vec.end()){
            it->second.push_back(info_[i]);
        }else{
            vector<Info> tmp;
            tmp.push_back(info_[i]);
            vec.insert(pair<int, vector<Info> >(info_[i].group_, tmp));
        }
    }
}

void DBscan::setInfo(vector<Info> info){
    info_ = info;
}
void DBscan::judge_minpts(double radius, vector<Info> pip, Info r, vector<Info> &N){
    int count_tmp = 0;
    for(int i = 0; i < pip.size(); i++){
double dis = earth_distance(r.x_, r.y_, pip[i].x_, pip[i].y_);
        if(dis < radius && dis != 0.0){
            count_tmp++;
            N.push_back(pip[i]);
        }
    }
}
void DBscan::do_dbscan(int minpts, double radius){
    vector<Info> pip = info_;
    info_.clear();
    int group = 1;
    while(pip.size() > minpts){
        vector<Info> N;
        int r = rand() % pip.size();
        pip[r].group_ = group;
        pip[r].visited_ = true;
        N.push_back(pip[r]);
        judge_minpts(radius, pip, pip[r], N);
        std::cout << "########## pip: " << pip.size() << std::endl;
        std::cout << "########## N: " << N.size() << std::endl;
        std::cout << "######### group: " << group << std::endl;
        if(N.size() > minpts){
            std::cout << "%%%%%%%%%%%%%%%%%%%%::" << N.size() << std::endl;
            for(int i = 0; i < N.size(); i++){
                if(N[i].visited_)
                    continue;
                N[i].visited_ = true;
                vector<Info> N1;
                judge_minpts(radius, pip, N[i], N1);
                std::cout << "###########judge:" << N1.size() << " N:" << N.size() << std::endl;
                if(N1.size() > minpts){
                    for(int j = 0; j < N1.size(); j++){
                        bool flag_has = false;
                        for(int k = 0; k < N.size(); k++){
                            if(N[k].id_ == N1[j].id_){
flag_has = true;
                                break;
                            }
                        }
                        if(flag_has){
                            continue;
                        }
                        N.push_back(N1[j]);
                    }
                }
                if(N[i].group_ == -1){
                    N[i].group_ = group;
                }
            }
        }else{
            pip[r].group_ = -2;
            for(vector<Info>::iterator it = pip.begin(); it != pip.end();){
                if(it->id_ == pip[r].id_){
                    pip.erase(it);
                    break;
                }else{
                    it++;
                }
            }
            continue;
        }

        group++;
        for(int i = 0; i < N.size(); i++){
            info_.push_back(N[i]);

            for(vector<Info>::iterator it = pip.begin(); it != pip.end();){
                if(it->id_ == N[i].id_ && it != pip.end()){
                    pip.erase(it);
                    break;
}else if(it != pip.end()){
                    it++;
                }
            }
        }
        std::cout << "!!!!!!!!!!!!!!info: " << info_.size() << std::endl;
        std::cout << "!!!!!!!!!!!!!!pip: " << pip.size() << std::endl;
   }
}

double DBscan::earth_distance(double lat1, double lng1, double lat2, double lng2){
    double radLat1 = deg2rad(lat1);
    double radLat2 = deg2rad(lat2);
    double a = radLat1 - radLat2;
    double b = deg2rad(lng1) - deg2rad(lng2);

    double dst = 2.0 * asin( (sqrt( pow(sin(a / 2), 2) + cos(radLat1) * cos(radLat2) * pow(sin(b / 2), 2))));

    dst = dst * 6378.137;
    return dst;
}

double DBscan::deg2rad(double d){
    return d * 3.1415926535898 / 180.0;
}

DBscan getData(){
    string input_file = "poi_beijing";
    ifstream input(input_file.c_str());
    string line;
    DBscan data;
    vector<Info> info;
    int id = 0;
    while(getline(input, line)){
vector<string> set;
        boost::split(set, line, boost::is_any_of("\t"));
        if(set.size() >= 6){
            Info t;
            string value = set[0] + "\t" + set[3] + "\t" + set[4] + "\t" + set[5];
            vector<string> set1;
            boost::split(set1, set[5], boost::is_any_of("|"));
            if(set1.size() != 2)
                continue;
            boost::trim(set1[0]);
            boost::trim(set1[1]);
            //double x = lexical_cast<double>(set1[1]);
            //double y = lexical_cast<double>(set1[0]);
            stringstream ss;
            double x = 0, y = 0;
            ss << set1[1];
            ss >> x;
            ss.clear();
            ss << set1[0];
            ss >> y;
            t.setData(id, -1, value, false, x, y);
            id++;
            std::cout << "id: " << id  << ", x: " << t.x_ << ", y: " << t.y_ << std::endl;
            info.push_back(t);
        }
    }
    data.setInfo(info);
    input.close();
    return data;
}

void cluster_to_file(map<int, vector<Info> > vec, string tar)
{
    ofstream output(tar.c_str());
    for(map<int, vector<Info> >::iterator it = vec.begin(); it != vec.end(); it++){
output << "group\t" << it->first << "\n";
        for(int i = 0; i < it->second.size(); i++){
            output << "id\t" << it->second[i].id_ << "\tvalue\t" << it->second[i].value_ << "\n";
        }
        output << "\n";
    }
    output.close();
}


int main()
{
    DBscan db = getData();
    db.do_dbscan(1000, 0.5);
    map<int, vector<Info> > vec;
    db.grouping(vec);
    cluster_to_file(vec, "test");
    return 0;
}

File: dbscan.py
import os
import sys
import math
import random
import re

default_encoding = "utf-8"

PI = 3.1415926535898
EARTH_RADIUS = 6378.137
if sys.getdefaultencoding() != default_encoding:
    reload(sys)
    sys.setdefaultencoding(default_encoding)

def deg2rad(d):
    return d * PI / 180.0

def earth_distance(lat1, lng1, lat2, lng2):
    radLat1 = deg2rad(lat1)
    radLat2 = deg2rad(lat2)
    a = radLat1 - radLat2
    b = deg2rad(lng1) - deg2rad(lng2)

    dst = 2.0 * math.asin((math.sqrt(math.pow(math.sin(a / 2), 2) + math.cos(radLat1) * math.cos(radLat2) * math.pow(math.sin(b / 2), 2))))
    dst = dst * EARTH_RADIUS
    return dst

class Info:
    id_ = 0
    group_ = 0
    value_ = ""
    visited_ = 0
    x_ = 0.0
    y_ = 0.0
    def __init__(self):
        self.id_ = 0
    def setData(self, id_, group_, value_, visited_, x_, y_):
        self.id_ = id_
        self.group_ = group_
        self.value_ = value_
        self.visited_ = visited_
        self.x_ = x_
        self.y_ = y_
    def print_info(self):
        print "id_: " + str(self.id_) + ", group_: " + str(self.group_) + ", value_: " + self.value_ + ", visited: " + str(self.visited_) + ", x_: " + str(self.x_) + ", y_: "+ str(self.y_)

class DBscan:
    info_ = []
    def __init__(self):
        self.info_ = []
    def grouping(self, vec):
        for i in xrange(0, len(self.info_)):
            if vec.has_key(self.info_[i].group_):
                vec[self.info_[i].group_].append(self.info_[i])
            else:
                tmp = []
                tmp.append(self.info_[i])
                vec[self.info_[i].group_] = tmp
        return vec
    def setInfo(self, info_):
        self.info_ = info_
    def judge_minpts(self, radius, pip, r, N):
        for i in xrange(0, len(pip)):
            dis = earth_distance(r.x_, r.y_, pip[i].x_, pip[i].y_)
            if dis < radius and dis != 0.0:
                N.append(pip[i])
        return N
    def do_dbscan(self, minpts, radius):
        pip = self.info_
        newinfo = []
        group = 1
        while len(pip) >= minpts:
            N = []
            r = random.randint(0, len(pip) - 1)
            pip[r].group_ = group
            pip[r].visited_ = 1
            N.append(pip[r])
            N = self.judge_minpts(radius, pip, pip[r], N)
            if len(N) >= minpts:
                i = 0
                while i < len(N):
 if N[i].visited_ == 1:
                        i = i + 1
                        continue
                    N[i].visited_ = 1
                    N1 = []
                    N1 = self.judge_minpts(radius, pip, N[i], N1)
                    if len(N1) >= minpts:
                        j = 0
                        while j < len(N1):
                            flag_has = 0
                            k = 0
                            while k < len(N):
                                if N[k].id_ == N1[j].id_:
                                    flag_has = 1
                                    break
                                k = k + 1
                            if flag_has == 1:
                                j = j + 1
                                continue
                            N.append(N1[j])
                            j = j + 1
                    if N[i].group_ == 0:
                        N[i].group_ = group
                    i = i + 1
            else:
                del pip[r]
                continue
            group = group + 1
            for i in xrange(0, len(N)):
                newinfo.append(N[i])
                for j in xrange(0, len(pip)):
                    if pip[j].id_ == N[i].id_:
                        del pip[j]
                        break
        self.info_ = newinfo
def getData(vec):
    data = DBscan()
    info = []
    ID = 0
    for i in xrange(0, len(vec)):
        t = Info()
        tmp_set = vec[i].split("\t")
        xys = tmp_set[4].split(",")
        if len(xys) != 2:
            continue
        x = float(xys[0])
        y = float(xys[1])
        t.setData(ID, 0, vec[i], 0, x, y)
        ID = ID + 1
        info.append(t)
    data.setInfo(info)
    return data

def judge_branch_school(line):
    line = line.strip().strip("\t").strip("\n")
    tmp_set = line.split("\t")
    if len(tmp_set) < 3 or line.find("培训机构") != -1 or line.find("驾校") != -1 or line.find("媒体") != -1:
        return 0
    sub_vec = []
    sub_vec.append("校区")
    sub_vec.append("分校")
    sub_vec.append("大学")
    for i in xrange(0, len(sub_vec)):
        m1 = re.search('.*' + sub_vec[i] + '$', tmp_set[1])
        m2 = re.search('.*' + sub_vec[i] + '\)$', tmp_set[1])
        m3 = re.search('.*' + sub_vec[i] + '）$', tmp_set[1])
        if m1 or m2 or m3:
            return 1
    return 0

if __name__ == "__main__":
    input_file = open("college_by_city", "r")
    vec = []
    for line in input_file:
        line = line.strip().strip("\t").strip("\n")
        tmp_set = line.split("\t")
        if len(tmp_set) >= 3:
            vec.append(line)
        elif len(tmp_set) < 3:
            if len(vec) <= 1:
                del vec[:]
                continue
            if len(vec) > 1:
                db = getData(vec)
                db.do_dbscan(2, 1)
                vec_dict = {}
                vec_dict = db.grouping(vec_dict)
                #print len(vec_dict)
                print_flag = 0
                if len(vec_dict) <= 1:
                    del vec[:]
                    continue
                #print len(vec_dict)
                for (k, v) in vec_dict.items():
                    #if len(v) <= 1:
                     #   continue
                    branch_school = []
                    for i in xrange(0, len(v)):
                        if judge_branch_school(v[i].value_) == 1: #以分校，校区，学校结尾的直接加入结果集
                            branch_school.append(v[i].value_)
                    if len(branch_school) == 0:
                        shortest = ""
                        shortest_flag = -1
                        sub_flag = 0
                        for j in xrange(0, len(v)):
shortest_tmp = v[j].value_.split("\t")
                            if v[j].value_.find("校区") != -1 or v[j].value_.find("分校") != -1:
                                shortest = shortest_tmp[1]
                                shortest_flag = j
                                sub_flag = 1
                                break
                        if sub_flag == 0:
                            for j in xrange(0, len(v)):
                                shortest_tmp = v[j].value_.split("\t")
                                if j == 0 and shortest == "":
                                    shortest = shortest_tmp[1]
                                    shortest_flag = j
                                else:
                                    if len(shortest_tmp[1]) < len(shortest):
                                        shortest = shortest_tmp[1]
                                        shortest_flag = j
                        else:
                            for j in xrange(0, len(v)):
                                shortest_tmp = v[j].value_.split("\t")
                                if (v[j].value_.find("校区") != -1 or v[j].value_.find("分校") != -1) and len(shortest_tmp[1]) < len(shortest):
                                    shortest = shortest_tmp[1]
                                    shortest_flag = j

                        if shortest_flag != -1:
                            branch_school.append(v[shortest_flag].value_)
                    if len(branch_school) != 0:
                        print_flag = 1
                        for j in xrange(0, len(branch_school)):
                            print branch_school[j]
                if print_flag == 1:
                    print "\n"

            del vec[:]
    input_file.close()
